File: android/app/src/main/java/com/mrt/app/core/models/FileModels.kt
package com.mrt.app.core.models

data class FileEntryModel(
    val name: String,
    val path: String,
    val isDirectory: Boolean,
    val size: Long,
    val modifiedMs: Long,
) {
    companion object {
        fun fromProto(entry: mrt.Mrt.FileEntry): FileEntryModel =
            FileEntryModel(
                name = entry.name,
                path = entry.path,
                isDirectory = entry.isDir,
                size = entry.size,
                modifiedMs = entry.modifiedMs,
            )
    }
}

enum class FilesUnavailableReason {
    DISCONNECTED,
    NO_ACTIVE_SESSION,
    SESSION_UNAVAILABLE,
}

sealed interface FileViewerState {
    data object None : FileViewerState
    data class Loading(val path: String) : FileViewerState
    data class Editable(
        val path: String,
        val content: String,
        val isSaving: Boolean,
        val errorMessage: String?,
    ) : FileViewerState
    data class ReadOnly(val path: String, val message: String) : FileViewerState
    data class Error(val path: String, val message: String) : FileViewerState
}

sealed interface FilesViewState {
    data class Unavailable(val reason: FilesUnavailableReason) : FilesViewState
    data class LoadingDirectory(val path: String) : FilesViewState
    data class DirectoryError(val path: String, val message: String) : FilesViewState
    data class DirectoryReady(
        val path: String,
        val entries: List<FileEntryModel>,
        val viewer: FileViewerState,
        val mutationMessage: String?,
    ) : FilesViewState
}

File: ios/MRT/Core/Models/FileModels.swift
import Foundation

struct FileEntryModel: Identifiable, Equatable {
    let name: String
    let path: String
    let isDirectory: Bool
    let size: UInt64
    let modifiedMs: UInt64

    var id: String { path.ifEmpty(name) }

    init(_ entry: Mrt_FileEntry) {
        self.name = entry.name
        self.path = entry.path
        self.isDirectory = entry.isDir
        self.size = entry.size
        self.modifiedMs = entry.modifiedMs
    }
}

enum FilesUnavailableReason: Equatable {
    case disconnected
    case noActiveSession
    case sessionUnavailable
}

enum FileViewerState: Equatable {
    case none
    case loading(path: String)
    case editable(path: String, content: String, isSaving: Bool, errorMessage: String?)
    case readOnly(path: String, message: String)
    case error(path: String, message: String)
}

enum FilesViewState: Equatable {
    case unavailable(FilesUnavailableReason)
    case loadingDirectory(path: String)
    case directoryError(path: String, message: String)
    case directoryReady(path: String, entries: [FileEntryModel], viewer: FileViewerState, mutationMessage: String?)
}

private extension String {
    func ifEmpty(_ fallback: String) -> String {
        isEmpty ? fallback : self
    }
}
